codelabs/Utopia Wholesale Codelab/Advanced Features/Deploying to Google Cloud Run.md

```diff
@@ -0,0 +1,279 @@
+   ```
+   - **Important**: Replace this with your own Cloud Run deployment URL after deploying the server
+   - Ensures proper redirects when accessing the root URL
+
+3. **Admin Password**: Added fixed `admin_password` set to `"multipaz-records"`
+   - Previously, the server generated a random password on each startup
+   - This makes the password predictable and manageable for deployment
+
+---
+
+## **Deployment Steps**
+
+### **Step 1: Create Dockerfile**
+
+Create a Dockerfile for the OpenID4VCI server in the `multipaz-openid4vci-server` directory:
+
+```bash
+mkdir -p multipaz-openid4vci-server && cat > multipaz-openid4vci-server/Dockerfile <<'EOF'
+FROM eclipse-temurin:17-jre
+WORKDIR /app
+COPY app.jar app.jar
+# Cloud Run sets PORT environment variable; pass it to the server using -param format
+CMD ["sh","-c","java -jar app.jar -param server_port=${PORT:-8080}"]
+EOF
+```
+
+**Explanation:**
+- `FROM eclipse-temurin:17-jre`: Uses Java 17 JRE base image (matches the server's Java version requirement)
+- `WORKDIR /app`: Sets the working directory inside the container
+- `COPY app.jar app.jar`: Copies the fat JAR file into the container
+- `CMD`: Runs the JAR with the PORT environment variable passed as a parameter
+  - Note: Uses `-param server_port=` format (not `--server.port=`) because this is a Ktor server, not Spring Boot
+  - `${PORT:-8080}` uses Cloud Run's PORT environment variable, defaulting to 8080 if not set
+
+### **Step 2: Build the Fat JAR**
+
+Build the fat JAR file that includes all dependencies:
+
+```bash
+./gradlew :multipaz-openid4vci-server:buildFatJar
+```
+
+**What this does:**
+- Compiles the Kotlin code
+- Packages all dependencies into a single JAR file
+- Includes the configuration files (including `default_configuration.json`)
+- Outputs the JAR to `multipaz-openid4vci-server/build/libs/app.jar`
+
+**Note:** The Ktor Gradle plugin creates a fat JAR named `app.jar` by default.
+
+### **Step 3: Prepare for Cloud Build**
+
+Copy the JAR file to the deployment directory:
+
+```bash
+# Rename and copy the JAR file to the multipaz-openid4vci-server directory
+cp multipaz-openid4vci-server/build/libs/app.jar multipaz-openid4vci-server/app.jar
+```
+
+**Directory structure should look like:**
+```
+multipaz-openid4vci-server/
+├── Dockerfile
+└── app.jar
+```
+
+### **Step 4: Build Docker Image**
+
+Build the Docker image using Google Cloud Build:
+
+```bash
+gcloud builds submit multipaz-openid4vci-server --tag gcr.io/engaged-list-481123-n3/multipaz-openid4vci-server
+```
+
+**Explanation:**
+- `gcloud builds submit`: Submits a build to Google Cloud Build
+- `multipaz-openid4vci-server`: The directory containing the Dockerfile and JAR file
+- `--tag`: Tags the resulting image with the specified name
+- `gcr.io/engaged-list-481123-n3/`: Your GCP project's Container Registry path
+  - Replace `engaged-list-481123-n3` with your actual GCP project ID
```

Comment:
add a warning emoji or something similar to catch attention (important step right?)